lib/redisHandler.js

```diff
@@ -0,0 +1,231 @@
+            if (this.listenerCount(command) !== 0) {
+                this.emit(command, data, responseId);
+            }
+            else {
+                this.sendError(responseId, new Error(command + 'NOT SUPPORTED'));
+            }
+        });
+        this.resp.on('finish', () => {
+            this.log.debug(this.sockedId + ' Redis finish');
+        });
+
+        socket.on('data', (data) => {
+            if (!data) return;
+            this.log.silly(this.sockedId + ' New Redis request: ' + data.toString().replace(/[\r\n]*/g, '')); // TODO remove data logging (performance)
+            this.resp.write(data);
+        });
+
+        socket.on('error', (err) => {
+            this.log.error(this.sockedId + ' Redis error: ' + err);
+            this.socket.destroy();
+        });
+    }
+
+    /**
+     * Check if the response to a certain command can be send out directly or
+     * if it needs to wait till earlier responses are ready
+     * @param responseId ID of the response
+     * @param data Buffer to send out
+     * @private
+     */
+    _sendQueued(responseId, data) {
+        let idx = 0;
+        while (this.writeQueue.length && idx <= this.writeQueue.length && this.writeQueue[idx]) {
+            if (this.writeQueue[idx].id === responseId) {
+                this.writeQueue[idx].data = data;
+                if (idx > 0) break;
+            }
+            if (idx === 0 && this.writeQueue[idx].data !== false) {
+                const response = this.writeQueue.shift();
+                this.log.silly(this.sockedId + ' Redis response (' + response.id + '): ' + response.data.toString().replace(/[\r\n]+/g, '')); // TODO remove data logging (performance)
+                this._write(response.data);
+            }
+            else {
+                idx++;
+            }
+        }
+
+        if (idx > 0) {
+            this.log.silly(this.sockedId + ' Redis response (' + responseId + '): Response queued');
+        }
+    }
+
+    /**
+     * Really write out a response to the network connection
+     * @param data Buffer to send out
+     * @private
+     */
+    _write(data) {
+        this.socket.write(data);
+    }
+
+    /**
+     * Guard to make sure a response is valid to be send out
+     * @param responseId ID of the response
+     * @param data Buffer to send out
+     */
+    send(responseId, data) {
+        if (responseId === null) {
+            this.log.silly(this.sockedId + ' Redis response DIRECT: ' + data.toString().replace(/[\r\n]+/g, '')); // TODO remove data logging (performance)
+            return this._write(data);
+        }
+        if (!responseId) {
+            throw Error('Invalid implementation: no responseId provided!');
+        }
+        if (!data) {
+            this.log.error('Not able to write ' + JSON.stringify(data));
+            data = Resp.encodeError(new Error('INVALID RESPONSE: ' + JSON.stringify(data)));
+        }
+        this._sendQueued(responseId, data);
```

Comment:
I'd break the possible stack here by using `setImmediate(() => this._sendQueued(responseId, data));` so callers don't have to wait for the while loop to be executed completely.

Reply:
makes sense